server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -696,7 +701,15 @@ private void raisePhaseFailure(SearchPhaseExecutionException exception) {
      * @see #onShardResult(SearchPhaseResult, SearchShardIterator)
      */
     final void onPhaseDone() {  // as a tribute to @kimchy aka. finishHim()
-        executeNextPhase(this, getNextPhase(results, this));
+        final SearchPhase nextPhase = getNextPhase(results, this);
+        // From src files the next phase is never null, but from tests this is a possibility. Hence, making sure that
+        // tests pass, we need to do null check on next phase.
+        if (nextPhase != null) {
+
+            final PipelinedRequest pipelinedRequest = searchPipelineService.resolvePipeline(this.getRequest());
+            pipelinedRequest.transformSearchPhase(results, this, this.getName(), nextPhase.getName());
```

Comment:
The call to `resolvePipeline` will transform the `SearchRequest` again.

We should call `resolvePipeline` exactly once per search request.

Reply:
A better approach is to pass the `PipelinedRequest` to the `AbstractSearchAsyncAction` in place of the `SearchRequest`, I think. (The `getRequest()` method can still return the wrapped `SearchRequest`.)

Reply:
> 

I am not seeing a way to do that. I think we need to build something here. But can we do it incrementally.

Reply:
> A better approach is to pass the `PipelinedRequest` to the `AbstractSearchAsyncAction` in place of the `SearchRequest`, I think. (The `getRequest()` method can still return the wrapped `SearchRequest`.)

Let me fix it.